### PR TITLE
Example of DID as top level element

### PR DESCRIPTION
**Pull request by faze-quadrata**
## Checklist
1. [ ] Unit Testing?
2. [ ] Backward Compatible?
3. [ ] Upgrade Safe (storage)?
4. [ ] Updated Confluence?
5. [ ] Updated Gitbook?

### CODE REVIEW

File: contracts/QuadPassport.sol [hunk 1]
```diff
@@ -74,11 +74,39 @@ contract QuadPassport is IQuadPassport, UUPSUpgradeable, QuadSoulbound, QuadPass
         bytes calldata _sigIssuer
     ) internal {
         address issuer = _setAttributesVerify(_account, _config, _sigIssuer);
+        bytes32 _attrKey;
+        bytes32 _attrType;
+
+        // Handle DID
+        if(_config.did != bytes32(0)){
```
Comment (fabrice-quadrata) [inline contracts/QuadPassport.sol:81]: Shouldn't you proceed to do this either way?

File: contracts/QuadPassport.sol [hunk 2]
```diff
@@ -74,11 +74,39 @@ contract QuadPassport is IQuadPassport, UUPSUpgradeable, QuadSoulbound, QuadPass
...
+            if (issuerPosition == 0) {
+            // Means the issuer hasn't yet attested to that attribute type
+                _attributes[_attrKey].push(attr);
+                _position[keccak256(abi.encode(_attrKey, issuer))] = _attributes[_attrKey].length;
+            } else {
+                // Issuer already attested to that attribute - override
+                _attributes[_attrKey][issuerPosition] = attr;
+            }
+
+        }
```
Comment (faze-quadrata) [inline contracts/QuadPassport.sol:101]: could probably be refactored later but this is just an example

File: contracts/QuadPassport.sol [hunk 3]
```diff
@@ -74,40 +74,58 @@ contract QuadPassport is IQuadPassport, UUPSUpgradeable, QuadSoulbound, QuadPass
...
         }
 
         if(balanceOf(_account, _config.tokenId) == 0)
             _mint(_account, _config.tokenId, 1);
         emit SetAttributeReceipt(_account, issuer, msg.value);
     }
 
+    function _writeAttrToStorage(
+        bytes32 _attrKey,
+        bytes32 _attrType,
```
Comment (fabrice-quadrata) [inline contracts/QuadPassport.sol:101]: not needed

File: contracts/QuadPassport.sol [hunk 4]
```diff
@@ -74,40 +74,58 @@ contract QuadPassport is IQuadPassport, UUPSUpgradeable, QuadSoulbound, QuadPass
...
+            _verifyAttrKey(_account, _config.attrTypes[i], _config.attrKeys[i], _config.did);
+            _writeAttrToStorage(_config.attrKeys[i], _config.attrTypes[i], _config.attrValues[i], issuer, _config.verifiedAt);
         }
 
         if(balanceOf(_account, _config.tokenId) == 0)
             _mint(_account, _config.tokenId, 1);
         emit SetAttributeReceipt(_account, issuer, msg.value);
     }
 
+    function _writeAttrToStorage(
```
Comment (fabrice-quadrata) [inline contracts/QuadPassport.sol:99]: can we add a quick docstring

**Review by fabrice-quadrata — APPROVED** [review]

**Pull request merged**